core/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java

```diff
@@ -0,0 +1,217 @@
+            Map<String, String> commitData = commits.get(commits.size() - 1).getUserData();
+            String translogGeneration = commitData.get(Translog.TRANSLOG_GENERATION_KEY);
+            String translogUUID = commitData.get(Translog.TRANSLOG_UUID_KEY);
+            if (translogGeneration == null || translogUUID == null) {
+                throw new ElasticsearchException("shard must have a valid translog generation and UUID but got: [{}] and: [{}]",
+                        translogGeneration, translogUUID);
+            }
+            terminal.println("Translog Generation: " + translogGeneration);
+            terminal.println("Translog UUID      : " + translogUUID);
+
+            Path tempEmptyCheckpoint = translogPath.resolve("temp-" + Translog.CHECKPOINT_FILE_NAME);
+            Path realEmptyCheckpoint = translogPath.resolve(Translog.CHECKPOINT_FILE_NAME);
+            Path tempEmptyTranslog = translogPath.resolve("temp-" + Translog.TRANSLOG_FILE_PREFIX +
+                            translogGeneration + Translog.TRANSLOG_FILE_SUFFIX);
+            Path realEmptyTranslog = translogPath.resolve(Translog.TRANSLOG_FILE_PREFIX + translogGeneration + Translog.TRANSLOG_FILE_SUFFIX);
+
+            // Write empty checkpoint and translog to empty files
+            long gen = Long.parseLong(translogGeneration);
+            writeEmptyCheckpoint(tempEmptyCheckpoint, gen);
+            writeEmptyTranslog(tempEmptyTranslog, translogUUID);
+
+            terminal.println("Removing existing translog files");
+            IOUtils.rm(translogFiles.toArray(new Path[]{}));
+
+            terminal.println("Creating new empty checkpoint at [" + realEmptyCheckpoint + "]");
+            Files.move(tempEmptyCheckpoint, realEmptyCheckpoint, StandardCopyOption.ATOMIC_MOVE);
+            terminal.println("Creating new empty translog at [" + realEmptyTranslog + "]");
+            Files.move(tempEmptyTranslog, realEmptyTranslog, StandardCopyOption.ATOMIC_MOVE);
+
+        } catch (LockObtainFailedException lofe) {
+            throw new ElasticsearchException("Failed to lock shard's directory at [" + idxLocation + "], is Elasticsearch still running?");
+        }
+
+        terminal.println("Done.");
+    }
+
+    /** Coerce a long into a byte array */
+    private static byte[] getLongBytes(long val) {
+        return new byte[] { (byte) (val >> 56), (byte) (val >> 48), (byte) (val >> 40), (byte) (val >> 32),
+                            (byte) (val >> 24), (byte) (val >> 16), (byte) (val >> 8), (byte) val };
+    }
+
+    /** Write a checkpoint file to the given location with the given generation */
+    public static void writeEmptyCheckpoint(Path filename, long translogGeneration) throws IOException {
+        try (OutputStreamDataOutput out = new OutputStreamDataOutput(new FileOutputStream(filename.toFile()))) {
+            Checkpoint emptyCheckpoint = new Checkpoint(43, 0, translogGeneration);
```

Comment:
Maybe add a comment explaining `43`?  Or, better, maybe derive it programatically, e.g. have `writeEmptyTranslog` return the number of bytes it wrote?